AStar multiple start vertices refactor

gmellemstrand opened the pull request:
To be completed by pull request submitter:

- [x] **issue**: [#2813 ](https://github.com/opentripplanner/OpenTripPlanner/issues/2813)
- [x] **roadmap**: [#2626 ](https://github.com/opentripplanner/OpenTripPlanner/issues/2626)
- [x] **tests**: I have deleted many of the GenericLocationTests and moved the ones that are still relevant to the LocationMatcher
- [x] **formatting**: Have you followed the [suggested code style](https://github.com/opentripplanner/OpenTripPlanner/blob/master/docs/Developers-Guide.md#code-style)? 
- [x] **documentation**: Added javadoc to the refactored GenericLocation class
- [x] **changelog**: add a bullet point to the [changelog file](https://github.com/opentripplanner/OpenTripPlanner/blob/master/docs/Changelog.md) with description and link to the linked issue

To be completed by @opentripplanner/plc:

- [ ] reviews and approvals by 2 members, ideally from different organizations
- [ ] **after merging**: update the relevant card on the [roadmap](https://github.com/orgs/opentripplanner/projects/1)

This is a refactor that is needed to support NeTEx extensions to stops and stations, but it should also make the process of specifying a from/to place from the API and resolving them to vertices cleaner.

This builds on [#2840 ](https://github.com/opentripplanner/OpenTripPlanner/pull/2840) and [#2868 ](https://github.com/opentripplanner/OpenTripPlanner/pull/2868)

What I have done:
- Simplified the main RoutingContext constructor. Removed transit-related logic and moved some logic out in separate methods.
- Changed the RoutingContext to contain a set of to/from vertices instead of a single one of each. This is needed because there is no edge between Station and Stop anymore.
- Removed origin/destination fields from RoutingContext, as they were used for reverse optimization.
- Changed the A Star implementation so that it can start from multiple initial states (one per start vertex).
- Refactored and simplified the GenericLocation. It now only contains a (pass-through) label, a FeedScopeId and coordinates. All pattern matching has been moved to a new LocationMatcher class, as this is only used in the REST API.
- Changed the getVerticesForLocation method in StreetVertexIndexServiceImpl to resolve the GenericLocation to a set of vertices by first trying to match the FeedScopeId to a Stop or StopCollection, and then matching the coordinate (via SimpleStreetSplitter) as a fallback.

abyrd commented:
@gmellemstrand I made most of the changes mentioned in my last comment-review, and in b0a287a also fixed a problem where we were repeatedly searching for the non-transit itinerary due to the old trip-banning loop. This should improve response times somewhat.

gmellemstrand commented:
@abyrd The changes look good to me. I have not reviewed the exact functionality of the LocationStringParser or the FareService implementation, but it does not affect routing otherwise.

I added another change to the NearbyStopFinder, to ensure that inital stop vertices are also added as StopForDistance objects.

gmellemstrand commented:
This fixes [#2770 ](https://github.com/opentripplanner/OpenTripPlanner/issues/2770)

gmellemstrand commented:
@abyrd You asked me about the NearbyStopFinder also including the origin vertices. This seems to already be accounted for before this pull request.

The NearbyStopFinder does include the origin vertices. This is what we need for the AccessEgressRouter. The DirectTransferGenerator, however does not want them to be included. This is accounted for by this check.

https://github.com/entur/OpenTripPlanner/blob/rutebanken_develop/src/main/java/org/opentripplanner/graph_builder/module/DirectTransferGenerator.java#L101
